rust/main/agents/relayer/src/server/messages/insert_messages.rs

```diff
@@ -0,0 +1,372 @@
+use std::collections::HashMap;
+
+use axum::{extract::State, http::StatusCode, routing::post, Json, Router};
+use derive_new::new;
+use serde::{Deserialize, Serialize};
+
+use hyperlane_base::{
+    db::{HyperlaneDb, HyperlaneRocksDB},
+    server::utils::{ServerErrorBody, ServerErrorResponse, ServerResult, ServerSuccessResponse},
+};
+use hyperlane_core::HyperlaneMessage;
+
+#[derive(Clone, Debug, new)]
+pub struct ServerState {
+    pub dbs: HashMap<u32, HyperlaneRocksDB>,
+}
+
+impl ServerState {
+    pub fn router(self) -> Router {
+        Router::new()
+            .route("/messages", post(handler))
+            .with_state(self)
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Message {
+    pub message: HyperlaneMessage,
+    pub dispatched_block_number: u64,
+}
+
+/// Request Body
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct RequestBody {
+    pub messages: Vec<Message>,
+}
+
+/// Response Body
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct ResponseBody {
+    pub count: u64,
+    pub skipped: Vec<Message>,
+}
+
+/// Manually insert messages into the database
+pub async fn handler(
+    State(state): State<ServerState>,
+    Json(payload): Json<RequestBody>,
+) -> ServerResult<ServerSuccessResponse<ResponseBody>> {
+    let RequestBody { messages } = payload;
+
+    let mut insertion_count: u64 = 0;
+    let mut skipped = Vec::new();
+    for message in messages {
+        tracing::debug!(?message, "Manually inserting message");
+
+        let dispatched_block_number = message.dispatched_block_number;
+        match state.dbs.get(&message.message.origin) {
+            Some(db) => {
+                store_message(db, &message.message, dispatched_block_number)?;
+                insertion_count += 1;
+            }
+            None => {
+                tracing::debug!(?message, "No db found for origin");
+                skipped.push(message);
+            }
+        }
+    }
+
+    let resp = ResponseBody {
+        count: insertion_count,
+        skipped,
+    };
+    Ok(ServerSuccessResponse::new(resp))
+}
+
+fn store_message(
+    db: &HyperlaneRocksDB,
+    message: &HyperlaneMessage,
+    dispatched_block_number: u64,
+) -> ServerResult<()> {
+    let id = message.id();
+    tracing::debug!(hyp_message=?message, "Storing new message in db");
+
+    db.store_message_by_id(&id, message).map_err(|err| {
```

Comment:
can we use the existing `store_message` https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/4d66b7328dbb0b519fe6f1d5c4db660f65f02921/rust/main/hyperlane-base/src/db/rocks/hyperlane_db.rs#L86 instead of replicating the implementation here? I wanna avoid a change there (like adding something new to the DB when storing a message) requiring another change here, which can be easy to miss. Imo it's not super necessary to have the individual error messages specific to where things failed

Reply:
Added :+1: